vendmachine/VendingMachine.java

```diff
@@ -0,0 +1,130 @@
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.io.FileWriter;
+import java.io.FileReader;
+import java.io.BufferedWriter;
+import java.io.BufferedReader;
+import java.io.IOException;
+
+class VendingMachine {
+    private HashMap<String, Product> products = new HashMap<>();
+    private BufferedWriter log = null;
+    private String user = null;
+
+    private void close() {
+        try {
+            BufferedWriter state = new BufferedWriter(new FileWriter("state.csv"));
+            Iterator it = products.values().iterator();
+            while (it.hasNext()) {
+                Product p = (Product) it.next();
+                state.write(p.getName()+", "+p.getPrice()+", "+p.getDescription()+"\n");
+            }
+            state.close();
+            log.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void log(String message) {
+        if (log == null) 
+            try {
+                log = new BufferedWriter(new FileWriter("transactions.log", true));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        try {
+            log.append(message+" "+(new Date())+"\n");
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void addProduct(String name, double price, String description) {
+        if (products.containsKey(name))
+            return;
+        Product p = new Product(name, price, description);
+        products.put(name, p);
+        log(user + " added product "+name+" "+price+" "+description);
+    }
+
+    public void changePrice(String name, double newprice) {
+        Product p = products.get(name);
+        if (p != null) {
+            log(user+" changed the price of "+name+" from "+p.getPrice()+" to "+newprice);
+            p.setPrice(newprice);
+        } else {
+            System.out.println("ERROR: attempting to change the price of the non-existing product "+name);
```

Comment:
💡 I would recommend throw an exception here with explanation rather than exposing it to stdout.